Firestore/Swift/Source/Codable/CollectionReference+WriteEncodable.swift

```diff
@@ -44,4 +44,36 @@ public extension CollectionReference {
     let encoded = try encoder.encode(value)
     return addDocument(data: encoded, completion: completion)
   }
+
+  /// Encodes an instance of `Encodable` and adds a new document to this collection
+  /// with the encoded data, assigning it a document ID automatically.
+  ///
+  /// See `Firestore.Encoder` for more details about the encoding process.
+  ///
+  /// - Parameters:
+  ///   - value: An instance of `Encodable` to be encoded to a document.
+  ///   - encoder: An encoder instance to use to run the encoding.
+  /// - Throws: `Error` if the backend rejected the write.
+  /// - Returns: A `DocumentReference` pointing to the newly created document.
+  @available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
+  @discardableResult
+  func addDocument<T: Encodable>(from value: T,
+                                 encoder: Firestore.Encoder = Firestore.Encoder()) async throws
+    -> DocumentReference {
+    return try await withCheckedThrowingContinuation { continuation in
+      var document: DocumentReference?
+      do {
+        document = try self.addDocument(from: value, encoder: encoder) { error in
+          if let error {
+            continuation.resume(throwing: error)
+          } else {
+            // Our callbacks guarantee that we either return an error or a document.
+            continuation.resume(returning: document!)
+          }
+        }
+      } catch {
```

Comment:
This catch looks redundant.

Reply:
It catches the possible thrown error by the original `addDocument` function at line 66. Since `withCheckedThrowingContinuation` takes a non-throwing `closure`, we can't just rethrow it.

Am I missing something here?

Reply:
Oh I see. It looks like the problem is with the existing API above that both throws and does a callback with an error parameter.

Reply:
Agree, I was about to refactor it but stopped because it is part of the public API and would be backward-incompatible. I can open another PR later for further discussion.

Reply:
About this PR,  how about moving this function back to the `CollectionReference+AsyncAwait` and using [the existing API there](https://github.com/firebase/firebase-ios-sdk/blob/26dab2702eb0d9756d5e149e275354b925fa55fe/Firestore/Swift/Source/AsyncAwait/CollectionReference%2BAsyncAwait.swift#L32) like:

```
  func addDocument<T: Encodable>(from value: T,
                                 encoder: Firestore.Encoder = Firestore.Encoder()) async throws
  -> DocumentReference {
    let encoded = try encoder.encode(value)
    return try await self.addDocument(data: encoded)
  }
```

So no need for `withCheckedThrowingContinuation` at all.

Reply:
See https://github.com/firebase/firebase-ios-sdk/issues/9157#issuecomment-1004637064 for a discussion about an edge case when the client is offline. This would result in the async method never returning.

Reply:
Using [this method](https://github.com/firebase/firebase-ios-sdk/pull/13407#discussion_r1694011572) will depend on the existing API. So, at the end of the day, we should either deprecate both of them or agree on the current existing logic.